### ISSUE TITLE
Feature request: in lib "theorems", make amount of hanging indent configurable

### ISSUE THREAD

**Issue by muzimuzhi** [T-F-S/tcolorbox#126]
Currently, a theorem environment defined by `\newtcbtheorem` will hanging indent the display name and separator in its title, which might not be users' expected behavior, see TeX-SX questions [1](https://tex.stackexchange.com/q/566094) and [2](https://tex.stackexchange.com/q/567137). Perhaps it's better to make the amount of hanging indent configurable.

A possible extension:

```tex
\documentclass{article}
\usepackage[theorems]{tcolorbox}
\usepackage{xpatch}

\makeatletter
% define key
%     hanging indent=<length> or "auto", (default and initially "auto")
% where "auto" means the width of display name and separator.
% If <length> is negative, this acts like \parindent=-<length>
\tcbset{
  hanging indent/.code={%
    \ifstrequal{#1}{auto}{%
      \let\tcb@theo@title@indent\tcb@theo@title@indent@std
    }{%
      \def\tcb@theo@title@indent##1{%
        \ifdim#1=0pt\relax
        \else
          \ifdim#1>0pt\relax
            \hangindent#1\hangafter=1\relax
          \else
            \hangindent\dimexpr-#1\relax\hangafter=-1\relax
          \fi
        \fi}
    }%
  },
  hanging indent/.default=auto
}

\def\tcb@theo@title@indent@std#1{\hangindent#1\hangafter=1\relax}

% init key and add it to reset style
\tcbset{
  hanging indent,
  reset@theorems/.append style=hanging indent
}

\xpatchcmd\tcb@theo@title
  {\hangindent\wd\z@\hangafter=1}
  {\tcb@theo@title@indent{\wd\z@}}
  {}{\fail}
\makeatother

\begin{document}
\newtcbtheorem{theorem}{Theorem}{}{theo}

\def\test#1{
  \begin{theorem}[#1, width=15em]{a slightly too long title}{test2}
    \ifstrempty{#1}{default setting}{\texttt{#1}}
  \end{theorem}}

\test{}
\test{hanging indent=auto}
\test{hanging indent=3em}
\test{hanging indent=0pt}
\test{hanging indent=-10pt}
\end{document}
```
![image](https://user-images.githubusercontent.com/6376638/96695525-540f4b80-13bc-11eb-86e6-12b648d5d35e.png)

**Comment by T-F-S** [comment]
Thank you for providing this feature request including the full code for implemented this feature!

I will incorporate this with some small adaptions for the next versions. The option name will be `theorem hanging indent`.

**Comment by T-F-S** [comment]
Implemented with version 4.50 (2021/05/21)

**Comment by muzimuzhi** [comment]
Just for the record, the related code changes are https://github.com/T-F-S/tcolorbox/commit/00f2d150b7152a7765fc3a4ba7c19b8368cc7283#diff-4df4173163ee68bb6f74c505be28f16753e400d7f965bf39a37cce9272741bc2.
